contracts/protocol/facets/ExchangeHandlerFacet.sol

```diff
@@ -519,8 +564,82 @@ contract ExchangeHandlerFacet is IBosonExchangeHandler, BuyerBase, DisputeBase {
         // Increase voucher counter for new buyer
         lookups.voucherCount[buyerId]++;
 
+        ProtocolLib.ProtocolStatus storage ps = protocolStatus();
+        if (ps.incomingVoucherCloneAddress != address(0)) {
+            ps.incomingVoucherId = _tokenId;
+        }
+
         // Notify watchers of state change
-        emit VoucherTransferred(exchange.offerId, _exchangeId, buyerId, msgSender());
+        emit VoucherTransferred(exchange.offerId, exchangeId, buyerId, msgSender());
+    }
+
+    /**
+     * @notice Handle pre-minted voucher transfer
+     *
+     * Reverts if
+     *
+     * @param _tokenId - the voucher id
+     * @param _to - the receiver address
+     * @param _from - the sender address
+     * @param _rangeOwner - the owner of the voucher range
+     * @param _sender - the caller address
+     */
+    function onPremintedVoucherTransferred(
+        uint256 _tokenId,
+        address payable _to,
+        address _from,
+        address _rangeOwner,
+        address _sender
+    ) external payable override buyersNotPaused returns (bool committed) {
```

Comment:
I believe reentrancy here is not problematic, since in general buyers shouldn't even have preminted vouchers to reenter.
But seller does, so we need to think if they can abuse it.

Maybe we can come up with a modified reentrancy guard (similar to what we have in metaTx, but a bit turned around), where `onPremintedVoucherTransferred` can be entered even if protocol was already entered, but we would limit it to only on "reenter".

Or if it safe the way it is now, it would be good to write it down to not forget it.